### DISCUSSION TITLE
Public IP missing as Candidate in SDP Answer?

### DISCUSSION THREAD

**Discussion by hqkirkland** [bluenviron/mediamtx#3603] [Questions] [answered]
### Question

Hey all,

Am I missing something? I'm trying to follow the guide on how to expose the WebRTC listener to a public address.

I've port-forwarded the UDP and TCP ports that I've selected for MediaMTX to listen on, but OBS isn't able to establish a connection. It should be noted that I can establish a raw TCP connection to the MediaMTX TCP listener using PuTTY (i.e. I confirmed traffic is not being intercepted by my router after portforwarding).

But the specified TCP and UDP ports are not offered for the public address I've specified. Only the public IP + UDP port.

WHIP Client: `68.21.x.x`
MediaMTX (v1.8.4) : `70.115.x.x`

`2024/08/03 13:08:29 INF [WebRTC] [session 73e97224] created by 68.21.x.x:51125
2024/08/03 13:08:29 DEB [path mystream] created
2024/08/03 13:08:29 DEB [WebRTC] [session 73e97224] peer connection state: connecting
2024/08/03 13:08:29 DEB [WebRTC] [conn 68.21.x.x:51125] [s->c] HTTP/1.1 201 Created
Accept-Patch: application/trickle-ice-sdpfrag
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: ETag, ID, Accept-Patch, Link, Location
Content-Type: application/sdp
Etag: *
Id: 73e97224-3356-4754-94b3-976ffd8f83cd
Link: <stun:stun.l.google.com:19302>; rel="ice-server"
Location: /mystream/whip/c387f927-d8af-4603-abde-c625c03de622
Server: mediamtx

(body of 2007 bytes)
2024/08/03 13:08:39 DEB [WebRTC] [session 73e97224] peer connection state: closed
2024/08/03 13:08:39 DEB [path mystream] destroyed: not in use
2024/08/03 13:08:39 INF [WebRTC] [session 73e97224] closed: deadline exceeded while waiting connection`

After studying the WHIP 201/Created response, I've discovered that the server is not offering the public address/port as a candidate in the SDP answer:
![image](https://github.com/user-attachments/assets/3f6859a9-dcee-4aa2-a3b6-0631ed722120)

You can see that the default Google STUN server is enabled, as well.

Btw: why do we need to configure the public IP ourselves? Shouldn't the STUN server yield the public address back to the server?

**Comment by hqkirkland** [comment]
Ok!

I found the answer! I think the README should be updated.

Instead of:
```
webrtcAdditionalHosts: [192.168.x.x, 1.2.3.4, my-dns.example.org, ...]
```

Use: 
```
# List of public IP addresses that are to be used as a host.
# This is used typically for servers that are behind 1:1 D-NAT.
webrtcICEHostNAT1To1IPs: [70.115.x.x]
```

@aler9 do I have this right? If so, I suggest updating the [README](https://github.com/bluenviron/mediamtx?tab=readme-ov-file#solving-webrtc-connectivity-issues) because I spent the whole day troubleshooting & trying to figure out why the right candidates weren't showing up, until I noticed this option in the YML.

**Reply by aler9** [reply]
Hello, this solution is not correct. `webrtcICEHostNAT1To1IPs` is a deprecated setting and is just an alias of `webrtcAdditionalHosts`. You just have to put your local IP address inside `webrtcAdditionalHosts`:

```
webrtcAdditionalHosts: [70.115.x.x]
```

> I suggest updating the [README](https://github.com/bluenviron/mediamtx?tab=readme-ov-file#solving-webrtc-connectivity-issues)

The README already says to fill `webrtcAdditionalHosts` with the public IP, in the first paragraph.

> Btw: why do we need to configure the public IP ourselves? Shouldn't the STUN server yield the public address back to the server?

For privacy reasons: in this way there's no need to contact an external server. If you prefer to use a STUN server rather than manually filling `webrtcAdditionalHosts`, you can do it by adding the STUN server inside the configuration file.

**Reply by hqkirkland** [reply]
Thanks @aler9  -- I tried `webrtcAdditionalHosts` as you suggested, and that works. I thought it was the other way around and it was `webrtcAdditionalHosts` being deprecated.

I understand WebRTC a lot better now and indeed, it's better to cut out STUN servers entirely if your server is a dedicated/static IP vs. dynamic.

**Comment by hqkirkland** [comment, marked as answer]
No need to update `README` -- users should use `webrtcAdditionalHosts` to specify their public IP as the README states.
